community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeCursor.java

```diff
@@ -157,7 +158,45 @@ public long allRelationshipsReference()
     @Override
     public long propertiesReference()
     {
-        return getNextProp();
+        //In the case where there hasn't been any changes in the transaction state this method simply returns the
+        // property reference.
+        //
+        //However if there has been changes we can have two cases:
+        //
+        //   i: The node had no prior properties, in this case we simply encode the node id in the reference
+        //      which allows the property cursor to probe tx state for properties.
+        //   ii: The node has properties, in this case we mark the actual property reference as having tx state
+        //
+        //in both cases we need to store a mapping from the computed reference to the state in the transaction state
+        //so we can retrieve it later in the property cursor.
```

Comment:
Yes you are right there is a theoretical chance this might happen if we acquire the property reference first, update the node and then use the stored property reference. If we would use the API like that we would miss updates. Currently we are not using it that way but we should fix it anyhow, I have created a card for moving the property reference magic to when we write instead of when we read.